apps/test/integration/levelSolutions/gamelab/timeout_api_blocks.js

```diff
@@ -0,0 +1,120 @@
+import tickWrapper from '../../util/tickWrapper';
+import {TestResults} from '@cdo/apps/constants';
+import {gamelabLevelDefinition} from '../../gamelabLevelDefinition';
+/* global Gamelab */
+
+/**
+ * @param {!string} testName
+ * @param {!string} program - Source code for the program to run during the
+ *   level test.
+ * @param {!function} doneCondition - A predicate used to check whether the
+ *   program is done running.
+ * @param {!function} validator - A set of assertions to run after the program
+ *   is complete.
+ * @returns {object} a level test definition.
+ */
+function testAsyncProgram(testName, program, doneCondition, validator) {
+  return {
+    description: testName,
+    editCode: true,
+    xml: program,
+    runBeforeClick: function (assert) {
+      // add a completion on timeout since this is a freeplay level
+      tickWrapper
+        .tickAppUntil(Gamelab, doneCondition.bind(null, assert))
+        .then(() => Gamelab.onPuzzleComplete());
+    },
+    customValidator: function (assert) {
+      validator(assert);
+      return true;
+    },
+    expected: {
+      result: true,
+      testResult: TestResults.FREE_PLAY
+    }
+  };
+}
+
+module.exports = {
+  app: "gamelab",
+  skinId: "gamelab",
+  levelDefinition: gamelabLevelDefinition,
+  tests: [
+    // These exercise the timeout API blocks
+    testAsyncProgram(
+      'setTimeout',
+      `
+        setTimeout(function() {
+          console.log('done');
+        }, 5);
+        
+        var key = setTimeout(function() {
+          console.log('do not expect this');
+        }, 4);
+        clearTimeout(key);
+      `,
+      function isProgramDone() {
+        const debugOutput = document.getElementById('debug-output').textContent;
+        return debugOutput.includes('done');
```

Comment:
I love this pattern! How does it work?

Reply:
what I mean is, where does `isProgramDone` get called from such that it's contents actually determines when to stop the program?

Reply:
I realized there was a lot of dupilcated boilerplate between my level test definitions, so I generated a helper (`testAsyncProgram` at the top of this file) that generates the level test definitions for me.  It's a nice way to encode some shared assumptions across a set of tests.

Here, the shared assumptions include:
- I'm writing a Game Lab test
- The test is async, and will need to run until some predicate is satisfied
- I'll be using a custom validator.

The generated level test populates the given JavaScript source code into the level, clicks "Run", uses the `tickAppUntil` utility to check the given `isProgramDone` predicate on every tick and then clicks "Finish" when it's satisfied, and then runs the `validateResult` custom validator.  Note the names of the functions here are just for readability, they could be anonymous functions.

I've found checking debug output is an easy way to verify interpreter behaviors in Game Lab tests - easier, anyway, than trying to assert something about the visualization.

Reply:
Ahh makes sense, I was lazily looking for other occurrences of `isProgramDone`, but I see now how this is being called by testAsyncProgram. Very nice!

Reply:
I considered making testAsyncProgram take a config object instead (effectively named arguments) - that might have made this code easier to trace.  I'll consider that for future cleanup.